pennylane/qaoa/layers.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Methods that define cost and mixer layers for use in QAOA workflows.
+"""
+import pennylane as qml
+from pennylane.operation import Tensor
+
+
+def _diagonal_terms(hamiltonian):
+    r"""Checks if all terms in a Hamiltonian are products of diagonal Pauli gates (PauliZ and Identity).
+
+    Args:
+        hamiltonian (.Hamiltonian): The Hamiltonian being checked
+
+    Returns:
+        bool: ``True`` if all terms are products of diagonal Pauli gates, ``False`` otherwise
+    """
+    val = True
+
+    for i in hamiltonian.ops:
+        i = Tensor(i) if isinstance(i.name, str) else i
+        for j in i.obs:
+            if j.name != "PauliZ" and j.name != "Identity":
+                val = False
+
+    return val
+
+
+def cost_layer(hamiltonian):
```

Comment:
I find it a bit strange at first sight that the layer does not take `gamma` as an input. I understand that we call ApproxTimeEvolution as a lambda function, but I'm wondering if this is the best choice. 

Alternatively we could use `def cost_layer(hamiltonian, gamma):` and then use
```
cost_layer = qaoa.cost_layer(cost_h, gamma)
@qml.qnode(dev)
    def circuit(gamma):
        cost_layer
```
right? Not sure which approach is best. Curious what others think.

Reply:
I am ambivalent 🤔 

I see pros and cons for both approaches:

* In the `cost_layer(cost_h, gamma)` approach, `cost_layer` is essentially a template, and must always be called within a QNode.

* In the `cost = cost_layer(cosh_h)` approach, `cost_layer` is more like a 'template generating function'. The returned cost function can only be used within a QNode, like a template.

Reply:
Probably lean more towards `cost_layer(cost_h, gamma)`, unless you see a use-case for having the 'template generating' behaviour.

Reply:
@ixfoduap @josh146 Sounds good, I'll make the change!